### PR TITLE
fix: swap max for non native assets

### PR DESCRIPTION
**Pull request by herbig**
## Description

Brings back the existing Max button in the `TradeAssetInput` component, for ERC20 tokens.  This allows users to easily fill the input field with their max balance.

This was originally removed to simplify the logic for calculating gas costs, but since ERC20 tokens are not used for paying native gas it is not an issue to bring it back for those.

## Pull Request Type

- [X] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes https://github.com/shapeshift/web/issues/5828

## Risk

This change adds a listener to the Max button, causing it to appear, and fills the input field with the total balance, for non-native (ERC20) tokens only.

This is low-risk, it is unlikely there are unintended consequences to this.  Users are generally familiar with the placement and function of a Max button, so should understand what it is doing.

### Engineering

1. View the swap (Trade/Bridge) feature at http://localhost:3000/#/trade
2. Confirm that the Max button appears when swapping _FROM_ an ERC20 token to anything else.
3. Confirm that clicking the Max button inputs the value displayed in the top right (Balance) into the input field.
4. Confirm that the Max button DOES NOT appear when swapping from a native currency (e.g. ETH).

### Operations

This is user facing. Testing can be the same as above.

## Screenshots (if applicable)

Screenshot 2023-12-21 at 2 25 58 PM

### CODE REVIEW

**Review by 0xApotheosis — COMMENTED** [review]
Solid start. A few initial thoughts:

- We'll want to use the `ASSET_NAMESPACE` constant from the `caip` package, rather than the `erc20` string literal.
- Whilst the ticket AC does indeed say we only want to handle `erc20`s, we'll also likely want to handle `bep20`s (Binance Smart Chain), too. You'll notice non-native BSC assets do not currently support send max in this branch.
- We might want to look at using the `isEvmChainId` and `isNativeEvmAsset` helpers to achieve this in a more maintainable way.
- A few lint issues to fix up 🙏

**Comment by woodenfurniture** [comment]
Hey @herbig thanks for the PR! 🍻

File: src/components/MultiHopTrade/components/TradeAssetInput.tsx [hunk 1]
```diff
@@ -50,13 +52,16 @@ const AssetInputWithAsset: React.FC<AssetInputLoadedProps> = props => {
     if (props.onChange) props.onChange(balance, false)
   }, [balance, props])
 
-  const { assetNamespace } = fromAssetId(assetId)
+  const showMax = useMemo(
+    () => isEvmChainId(fromAssetId(assetId).chainId) && !isNativeEvmAsset(assetId),
+    [assetId],
+  )
```
Comment (0xApotheosis) [inline src/components/MultiHopTrade/components/TradeAssetInput.tsx:58]: Golden.

Nitpick, re-name to `shouldShowMax` to disambiguate that this is not an action.

**Review by gomesalexandre — COMMENTED** [review]
Solid start indeed, thanks a lot for the contribution @herbig! 

Following on the great review from @0xApotheosis, I've added a few comments, as this implementation is sane for the most part at runtime but potentially brings some major type safety risks.

From a functional perspective, the max button is indeed available for ERC-20 sell assets:

image

It looks like the scientific notation is used however, versus precision in the send modal, see:

image
image

You can see the send modal implementation for reference `BigNumber.toPrecision()` is what you want here:

https://github.com/shapeshift/web/blob/0f05cf8ef304d571586915ea871ca0486ee4bbfe/src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx#L199

Fiat max. is also happy.

image

And confirmed the max. button is only shown for EVM tokens.

File: src/components/MultiHopTrade/components/TradeAssetInput.tsx [hunk 2]
```diff
@@ -46,7 +48,23 @@ const AssetInputWithAsset: React.FC<AssetInputLoadedProps> = props => {
...
 
-  return <TradeAmountInput balance={balance} fiatBalance={fiatBalance} {...props} />
+  const onMaxClick = useCallback(() => {
+    if (props.onChange) props.onChange(balance, false)
+  }, [balance, props])
+
+  const showMax = useMemo(
+    () => isEvmChainId(fromAssetId(assetId).chainId) && !isNativeEvmAsset(assetId),
+    [assetId],
+  )
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/components/MultiHopTrade/components/TradeAssetInput.tsx]: `<TradeAssetInput` itself is passed `percentOptions`. If you inspect consumers, eventually reaching `<TradeInput />`, you'll see this guy:
https://github.com/shapeshift/web/blob/d613a96992863193e25fda34903320252c742df0/src/components/MultiHopTrade/components/TradeInput/TradeInput.tsx#L105

This is the prop that is controling which percent options are being displayed - `showMax` here is effectively redundant, as we're passing 1 (max) as an option from the parent, but then have children programmatically (not) render it

File: src/components/MultiHopTrade/components/TradeAssetInput.tsx [hunk 3]
```diff
@@ -46,7 +48,23 @@ const AssetInputWithAsset: React.FC<AssetInputLoadedProps> = props => {
   )
   const fiatBalance = bnOrZero(balance).times(marketData.price).toString()
 
-  return <TradeAmountInput balance={balance} fiatBalance={fiatBalance} {...props} />
+  const onMaxClick = useCallback(() => {
+    if (props.onChange) props.onChange(balance, false)
+  }, [balance, props])
+
+  const showMax = useMemo(
+    () => isEvmChainId(fromAssetId(assetId).chainId) && !isNativeEvmAsset(assetId),
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/components/MultiHopTrade/components/TradeAssetInput.tsx:56]: Either or, that one is fine as-is as we most likely won't support tokens from non-EVM chains for trading in the future but we may. Would prefer it regardless as it reads nicer and is more declarative

**Comment by gomesalexandre** [comment]
> > It looks like the scientific notation is used however, versus precision in the send modal, see:
> 
> @gomesalexandre oh good catch, yeah I didn't see that since I only checked a few assets that would not have triggered scientific notation
> 
> maybe I'm missing something, but to me it seems both of those screenshots would be wrong.
> 
> Scientific notation in the `Swap` field isn't great UX, and I'll need to fix that.

The scientific notation should never be user-facing, we always want to have either full-precision when applicable, or trimmed decimal places depending on the screens and product spec. For things like sends and swapping, we want to ensure we display the *actual* amounts that users are going to send out of their wallet.

i.e when sending FOX, you can see that the balance (for *display* purposes only) has trimmed precision, but max. click actually inputs the full-precision amount, with 18 digits in the case of FOX:

image


> But in the Send screen it looks as though you're sending 4.414, which isn't accurate.

Not sure what do you mean here @herbig? Clicking max/percent options will replace the input with the prorated one, with full asset precision, i.e in the screenshot above `115.618374110339184543` FOX (not e.g 115.61 FOX)


> Also so we're clear, the Max button in `Send` is already in production, and unrelated to Max here in the Swap, so I think we have a bug there in production. I've just tried prod Max and got this:

That's correct as per our current flow, though definitely not the best UX: we do fees deduction (on native assets only obviously) so trying to send 0.0003 ETH would result in more fees than the actual balance, hence the negative value.

**Comment by 0xean** [comment]
@herbig - thanks for the PR and work on this issue. @gomesalexandre is going to take it over from here since we haven't heard from you in a bit.  

I would love to send some FOX tokens your way as a way to thank you for the work on this, if you want to DM me I can get that arranged.  Thanks again for your contributions.

**Comment by gomesalexandre** [comment]
Closing in favor of https://github.com/shapeshift/web/pull/5961

**Pull request closed without merging**